### ISSUE TITLE
Direct support for flattened/packed arrays

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#466]
# What is that, and why would we want it

Flattened/packed/unpadded arrays/tensors are tensors consisting of multiple sequences of different length, but without padding. I.e. instead of a tensor [B,T,D1...Dn], you would have a tensor [B',D1...Dn] where B' = sum(seq_len). They represent the same data (you are supposed to ignore the padded values anyway). This can also be generalized for multiple dynamic axes [B,T1,T2,...,D1...Dn] and seq_lens1, seq_lens2, ..

We already make use of packed arrays in RETURNN in some places internally. See usages of `flatten_with_seq_len_mask`.

A packed array can **save memory**, and potentially **speed up** some computations. While other computations might be slower though. And in any case, you often need special implemented ops.

For purely feed-forward operations (basically all layers in RETURNN with `recurrent=False` flag, e.g. `LinearLayer`) it should not matter at all and just be faster. Recurrent operations, convolutional operations, softmax-over-spatial, and many other things would need special implementations.

(E.g. CuDNN LSTM early only supported to work on packed arrays. And also the sequences must be sorted by sequence length, longest first. But then they extended that to support also padded arrays, and now they even deprecated the packed format. Not sure why. Maybe it was just too unconvenient to use, because padded arrays are really the standard. The packed arrays cannot even be used anymore in TensorFlow. In PyTorch on the other side, packed arrays are still the standard for RNNs.)

I specifically want to use this for a more memory-friendly RNN-T loss implementation. I.e. there it would have two dynamic length axis.

# Proposal

RETURNN should more directly support packed arrays, i.e. via `Data`. Layers could potentially return such data, and get such data as input.

# Details (to be clarified)

## Implementation/API in `Data`

How exactly should we represent it in `Data`?

`placeholder` as `tf.Tensor` is clear, that will be the packed array.

`shape`, `batch_shape` etc directly describe `placeholder`, i.e. there is no time-axis anymore but the (batch) shape is [B',D1...Dn].

I think `size_placeholder` should be consistent to `shape`, so that would not contain the sequence length. We should store them separately somehow/somewhere.

We might want to add a new attribute `batch_dim`, which references a `DimensionTag`, and extend the `DimensionTag` for the batch dim. Or maybe use a separate or new object `BatchDim`. That could also store the beam search information, so that we can easily differentiate between different kinds of batch dims. That object can store the sequence lengths.

There are other auxiliary information which is useful for specific native ops, e.g. offsets. We can also store them in this object and use them when needed.

## Usage, Explicit vs Implicit

So far, the idea of the format in `Data` was that the user doesn't need to care about that, and layers would automatically convert into their most efficient format (e.g. time-major for RNNs), and all layers should be (mostly) invariant to the specific format / axis order so that the user doesn't need to think about this and it would always be correct. (That is at least in theory. In practice, it sometimes doesn't work as nicely, or is used incorrectly by the user.)

I'm not sure if such automatic logic also works for such padding / packed arrays. If we would adopt the same logic, layers could pack automatically when it is safe, and then leave as-is, and other layers might unpack when needed, or leave as-is.

This might have some problems:

* It might break probably many things, and needs some work to fix all nasty edge cases. I'm not sure if it also breaks user configs.
* It might actually not be the most efficient solution. E.g. consider a chain of LSTM + FFNN blocks (not that uncommon actually). When the FFNN (`LinearLayer`) would pack the tensor, then the LSTM would unpack it again (because our implementation expects the padded tensor). So there is a lot of overhead due to the packing + unpacking all the time. The question is whether this is still cheaper than the computation on the padded frames, but this is unclear.

Some explicit usage on the other side would not break any existing configs. But this might also not be optimal:

* The configs would contain unnecessary implementation details. And maybe this becomes outdated at some point (maybe at some point, padded arrays are really obsolete, or need a different API, etc).

There could also be a global config flag to enable this experimental automatic feature, which is disabled by default. Then it would not break anything.

If we try to do it automatic, the question is also, how exactly should the logic be like. Just greedily like we currently choose time-major vs batch-major? I.e. all layers which can potentially use this will use this. (E.g. `LinearLayer`. Frame-wise losses like CE. Or layers with specific support for this.)

If we want to have it explicit, we probably should just extend the `FlattenBatchLayer`.

When combining two flattened/packed arrays with different sequence axes (e.g. source and target; padded [B,T1,D], [B,T2,D], to [B,T1,T2,D]), this needs also special logic. It gets in [B',D] and [B'',D] and should output [B''',D].

---

(Btw, I was not sure whether this should be in the discussions forum or here as an issue. See [here](https://github.com/rwth-i6/returnn/discussions/465) for discussion.)

**Comment by albertz** [comment]
Some work in progress in #467.

**Comment by albertz** [comment]
I totally forgot about *ragged tensors* ([guide](https://www.tensorflow.org/guide/ragged_tensor), [`tf.ragged` module](https://www.tensorflow.org/api_docs/python/tf/ragged), [`tf.RaggedTensor` class](https://www.tensorflow.org/api_docs/python/tf/RaggedTensor)), which is the official TF representation for packed arrays. 
I'm not sure yet how/whether we should use this, and how much, where exactly...
(`Data.placeholder` probably should stay a pure `tf.Tensor`, nothing else? Otherwise lots of code might break.)

I also realized (or rather ignored so far), also for packed arrays, the order of (virtual) dimensions is still relevant, e.g. batch-major vs time-major. E.g. CuDNN expects time-major packed arrays. This information about dimension order must be explicitly stored (e.g. in `BatchInfo`), and also we must have control over this order and support common operations (e.g. `transpose`).

I collected some thoughts (and open questions about optimal performance) [in this StackOverflow question](https://stackoverflow.com/questions/66533009/most-efficient-tensor-format-padded-vs-packed-time-major-vs-batch-major-for-r) (please vote to reopen this! I assume people just don't understand).

**Comment by albertz** [comment]
Note that this also potentially allows for static batch dims, although this is not fully tested yet (and there are likely some further bugs). But this could be interesting for TPU usage. This is kind of a necessary step towards TPU support.

**Comment by albertz** [comment]
I will stop working on this for now (or only make smaller fixes/extensions). It's already pretty useful as it is. Some future possible extensions:

* The batch size can in principle be static now. This is not well tested, but the support is there. This was an important building step for TPU support. (I think we also need static spatial axes. We could store that as a flag in the `DimensionTag`, sth like `spatial_static_max_size` or so, and then handle it in a similar way.) I think this mostly needs some testing and some smaller fixes, maybe a bit of work. (TPU support also must not use our native ops, but in many cases, we can just use standard TF variants instead. Or maybe we find also some way to compile native to XLA or so.)
* Flattening multiple spatial dimensions together into the batch dim is not implemented yet. This is the interesting case to really safe memory. E.g. for RNNT training. But this is a bit tricky. `copy_extend_batch` should be extended for this. And then we would also need special implementations for e.g. RNNT or any other operation which we would want to do on this. Basically anything which is not feed-forward needs a special implementation. This is kind of the deal breaker to make this universally useful.
* It's not totally reliable currently. The code is currently written in a way that it is relaxed about incorrect or missing information in `data.batch`, and has some heuristics to fix them up. We should slowly fix that, maybe introduce warnings, and then errors at some point.
* We can make more use of this. E.g. in `SplitBatchTimeLayer`, we can don't need the user to provide an explicit `base` but can just get the information from the `batch`.
* `data.beam` is redundant with `data.batch.beam`. `data.batch` is actually more flexible, as it could also contain multiple beams. This should be cleaned up. But for that, we must make sure that it is totally reliable.
 
I think it makes sense to close this issue, as the basic support is there now. For anything more advanced, let's make separate issues.
